apm-agent-core/src/main/java/co/elastic/apm/agent/impl/InScopeRunnableWrapper.java

```diff
@@ -0,0 +1,87 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 Elastic and contributors
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.impl;
+
+
+import co.elastic.apm.agent.bci.VisibleForAdvice;
+import co.elastic.apm.agent.impl.transaction.Transaction;
+import co.elastic.apm.agent.objectpool.Recyclable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+@VisibleForAdvice
+public class InScopeRunnableWrapper implements Runnable, Recyclable {
+    private final Logger logger = LoggerFactory.getLogger(InScopeRunnableWrapper.class);
+
+    @Nullable
+    private Runnable delegate;
+    @Nullable
+    private Transaction transaction;
+    @Nullable
+    private ElasticApmTracer tracer;
+
+    public InScopeRunnableWrapper( ElasticApmTracer tracer) {
+        this.tracer = tracer;
+    }
+
+    public InScopeRunnableWrapper wrap(Runnable delegate) {
+        this.delegate = delegate;
+        return this;
+    }
+
+    public InScopeRunnableWrapper withTransaction(Transaction transaction) {
+        this.transaction = transaction;
+        return this;
+    }
+
+    @Override
+    public void run() {
+        Scope scope = null;
```

Comment:
I think this can be simplified with
```
if (transaction != null) {
    transaction.activate();
    try {
        delegate.run();
    } catch (Exception e) {
        transaction.captureException(e);
        throw e;
    } finally {
        transaction.deactivate();
    }
}
```

Reply:
Well, I don't want to give up the try/catch around `activate`/`deactivate` - it's just like your other PR to catch errors during advices - our code should never prevent user code from running. I can give up the usage of scope though.

Your suggestion to capture the Exception and report as error is tempting, but is not safe here. Notice that this code is not taking responsibility on ending and reporting the Transaction, it's just responsible for scoping in that thread. For example, the usage of `AsyncContext.onComplete` would end the transaction in this PR (and `AsyncContext.onError` will report errors). So I don't know what the status of the transaction is at this stage.

Reply:
Good point on the problems with `captureException`. However, `AbstractSpan#activate` already catches exceptions.

Reply:
Is it valid to deactivate a finished span? Ideally, `#end` should be the last call on a span because after that the span is recycled and could be reused any time. Deactivating after that is borderline but probably unproblematic if the state isn't changed after `#end`.